lilypad/llm_fn.py

```diff
@@ -100,26 +154,76 @@ async def inner_async(*args: _P.args, **kwargs: _P.kwargs) -> _R:
             def inner(*args: _P.args, **kwargs: _P.kwargs) -> _R:
                 arg_types, arg_values = inspect_arguments(fn, *args, **kwargs)
                 version = get_llm_function_version(fn, arg_types, synced)
-                trace_decorator = trace(
-                    project_id=lilypad_client.project_id,
-                    version_id=version.id,
-                    arg_types=arg_types,
-                    arg_values=arg_values,
-                    lexical_closure=version.llm_fn.code,
-                    prompt_template=version.fn_params.prompt_template
-                    if version.fn_params
-                    else "",
-                    version=version.version,
-                )
+
+                @contextmanager
+                def custom_context_manager(
+                    fn: Callable,
+                ) -> Generator[Span, Any, None]:
+                    tracer = get_tracer("lilypad")
+                    with tracer.start_as_current_span(f"{fn.__name__}") as span:
+                        attributes: dict[str, AttributeValue] = {
+                            "lilypad.project_id": lilypad_client.project_id
+                            if lilypad_client.project_id
+                            else 0,
+                            "lilypad.function_name": fn.__name__,
+                            "lilypad.version": version.version
+                            if version.version
+                            else "",
+                            "lilypad.version_id": version.id,
+                            "lilypad.arg_types": json.dumps(arg_types),
+                            "lilypad.arg_values": json.dumps(arg_values),
+                            "lilypad.lexical_closure": version.llm_fn.code,
+                            "lilypad.prompt_template": version.fn_params.prompt_template
+                            if version.fn_params
+                            else "",
+                            "lilypad.is_async": False,
+                        }
+                        span.set_attributes(attributes)
+                        yield span
 
                 if not synced:
-                    return cast(_R, trace_decorator(fn)(*args, **kwargs))
+                    if hasattr(fn, "__mirascope_call__"):
+                        decorator = middleware_factory(
```

Comment:
i think we could structure this better so we only use middleware_factory once.

For example:

```python
is_mirascope_call = hasattr(fn, "__mirascope_call__")
if not synced and not is_mirascope_call:
    decorator = trace(...)
    return cast(_R, decorator(fn)(*args, **kwargs))
decorator = middleware_factory(...)
if not synced and is_mirascope_call:
    return cast(_R, decorator(fn)(*args, **kwargs))

if not version.fn_params:
    ...

return traced_synced_llm_function_constructor(...)...
```

WDYT?